### PR TITLE
[sonic-swss] : PFCWD recovery changes using DLR_INIT

### PR DESCRIPTION
**Pull request by vmittal-msft**
**What I did**
This is changing PFCWD recovery from SONIC based to SAI/SDK DLR_INIT way.

**Why I did it**
Existing SONIC scheme will not work for upcoming BRCM devices starting with TH3.

**How I verified it**
Verified this to be working fine on HW platform where the issue was seen. Added sonic-mgmt test case to handle this scenario.

### CODE REVIEW

**Comment by prsunny** [comment]
Plans to add test?

**Comment by neethajohn** [comment]
> Plans to add test?

The changes required for existing testcases will be captured in this PR. We just need to include Th3 in the platform check
https://github.com/Azure/sonic-mgmt/pull/5665

**Comment by neethajohn** [comment]
@vmittal-msft , as discussed, please explore if mock tests can be added for these changes

File: orchagent/pfcwdorch.cpp [hunk 1]
```diff
@@ -229,6 +232,36 @@ task_process_status PfcWdOrch<DropHandler, ForwardHandler>::createEntry(const st
...
+                                return task_process_status::task_invalid_entry;
+                            }
+
+                            setPfcDlrPacketAction(action);
+                        }
+                        else
+                        {
+                            if(checkPfcDlrPacketAction() != action) 
+                            {
+                                SWSS_LOG_ERROR("Invalid PFC Watchdog action %s as switch level action is set", value.c_str());
```
Comment (neethajohn) [inline orchagent/pfcwdorch.cpp:258]: include the switch level action also in this msg

File: orchagent/pfcwdorch.cpp [hunk 2]
```diff
@@ -229,6 +232,36 @@ task_process_status PfcWdOrch<DropHandler, ForwardHandler>::createEntry(const st
...
+                                return task_process_status::task_invalid_entry;
+                            }
+                        }
+
```
Comment (prsunny) [inline orchagent/pfcwdorch.cpp:262]: Remove extra line

File: orchagent/pfcwdorch.cpp [hunk 3]
```diff
@@ -229,6 +232,36 @@ task_process_status PfcWdOrch<DropHandler, ForwardHandler>::createEntry(const st
...
+                            attr.id = SAI_SWITCH_ATTR_PFC_DLR_PACKET_ACTION;
+                            attr.value.u32 = (sai_uint32_t)action;
+
+                            sai_status_t status = sai_switch_api->set_switch_attribute(gSwitchId, &attr);
+                            if(status != SAI_STATUS_SUCCESS)
+                            {
+                                SWSS_LOG_ERROR("Failed to set switch level PFC DLR packet action rv : %d", status);
+                                return task_process_status::task_invalid_entry;
+                            }
+
```
Comment (prsunny) [inline orchagent/pfcwdorch.cpp:251]: remove extra line

File: orchagent/switchorch.cpp [hunk 4]
```diff
@@ -59,6 +59,23 @@ SwitchOrch::SwitchOrch(DBConnector *db, vector<TableConnector>& connectors, Tabl
...
+        m_PfcDlrInitEnable = false;
+        fvVector.emplace_back(SWITCH_CAPABILITY_TABLE_PFC_DLR_INIT_CAPABLE, "false");
+    }
+    else 
+    {
+        SWSS_LOG_INFO("Queue level PFC DLR INIT configuration is supported");
+        m_PfcDlrInitEnable = true;
+        fvVector.emplace_back(SWITCH_CAPABILITY_TABLE_PFC_DLR_INIT_CAPABLE, "true");
+    }
+    set_switch_capability(fvVector);
```
[the remark this answers is not in the document]
Comment (vmittal-msft) [reply orchagent/switchorch.cpp:78]: Done.

File: tests/mock_tests/portsorch_ut.cpp [hunk 5]
```diff
@@ -20,6 +20,32 @@ namespace portsorch_test
...
+    sai_queue_api_t ut_sai_queue_api;
+    sai_queue_api_t *pold_sai_queue_api;
+    int _sai_set_queue_attr_count = 0;
+
+    sai_status_t _ut_stub_sai_set_queue_attribute(
+        _In_ sai_object_id_t queue_id,
+        _In_ uint32_t attr_count,
+        _Inout_ sai_attribute_t *attr_list)
+    {
+        _sai_set_queue_attr_count++;
```
Comment (neethajohn) [inline tests/mock_tests/portsorch_ut.cpp:32]: please check for the specific sai attribute and bool value and then increment count.

File: tests/mock_tests/portsorch_ut.cpp [hunk 6]
```diff
@@ -799,4 +825,99 @@ namespace portsorch_test
...
+        gBufferOrch->addExistingData(&queueTable);
+
+        // process pool, profile and PGs
+        static_cast<Orch *>(gBufferOrch)->doTask();
+
+        auto countersTable = make_shared<Table>(m_counters_db.get(), COUNTERS_TABLE);
+        auto dropHandler = make_unique<PfcWdDlrHandler>(port.m_port_id, port.m_queue_ids[3], 3, countersTable);
+
+        ASSERT_TRUE(_sai_set_queue_attr_count == 1);
+
```
Comment (neethajohn) [inline tests/mock_tests/portsorch_ut.cpp:920]: release the handler and verify that the sai attribute is set with bool false

Comment (vmittal-msft) [reply tests/mock_tests/portsorch_ut.cpp:1084]: Done

File: tests/mock_tests/portsorch_ut.cpp (hunk 5 above)
Comment (vmittal-msft) [reply tests/mock_tests/portsorch_ut.cpp]: Done

**Review by gechiang — DISMISSED** [review]
LGTM

**Review by prsunny — DISMISSED** [review]

**Review by neethajohn — DISMISSED** [review]

**Review by gechiang — APPROVED** [review]

**Pull request merged**